src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorage.cpp

```
@@ -290,6 +290,12 @@ void MetadataStorageFromPlainObjectStorageTransaction::createDirectory(const std
         return;
 
     auto normalized_path = normalizeDirectoryPath(path);
+    if (normalized_path.empty())
+    {
+        LOG_TRACE(getLogger("MetadataStorageFromPlainObjectStorageTransaction"), "Skipping creation of a directory with an empty path");
```

Should we throw an exception instead?

And can we print out the original `path`, so we know if something wrong with `normalizeDirectoryPath`?

---

Metadata objects corresponding to empty directories may already be written. Throwing an exception would mean we need to clean them manually.

---

>> Should we throw an exception instead?
> I essentially moved the check to src/Disks/DiskEncrypted.cpp

Reverting that, because it has some undesired side effects, e.g.:

tests/queries/0_stateless/03362_merge_tree_with_background_refresh.sh

```
[ip-172-31-8-86] 2025.08.30 03:18:34.873339 [ 1335461 ] {56d32d72-96e6-48eb-b4dd-7a0f6fd04002} <Error> executeQuery: Code: 281. DB::Exception: Directory path '' is empty. (DICTIONARY_IS_EMPTY) (version 25.9.1.1) (from [::ffff:127.0.0.1]:61872) (query 1, line 2) (in query: CREATE TABLE writer (`s` String) ORDER BY tuple() SETTINGS table_disk = true, disk = disk(name = `03362_writer_test`, type = object_storage, object_storage_type = '[HIDDEN]', metadata_type = '[HIDDEN]', path = '[HIDDEN]')), Stack trace (when copying this message, always include the lines below):

0. ./contrib/llvm-project/libcxx/include/__exception/exception.h:113: Poco::Exception::Exception(String const&, int) @ 0x000000001f7bce32
1. ./build.release/./src/Common/Exception.cpp:128: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000fad999e
2. DB::Exception::Exception(String&&, int, String, bool) @ 0x000000000904a38e
3. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000009049f80
4. DB::Exception::Exception<String const&>(int, FormatStringHelperImpl<std::type_identity<String const&>::type>, String const&) @ 0x000000000a11162b
5. ./build.release/./src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorage.cpp:297: DB::MetadataStorageFromPlainObjectStorageTransaction::createDirectory(String const&) @ 0x0000000015052943
6. ./build.release/./src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp:708: void std::__function::__policy_invoker<void (std::shared_ptr<DB::IMetadataTransaction>)>::__call_impl[abi:ne190107]<std::__function::__default_alloc_func<DB::DiskObjectStorageTransaction::createDirectories(String const&)::$_0, void (std::shared_ptr<DB::IMetadataTransaction>)>>(std::__function::__policy_storage const*, std::shared_ptr<DB::IMetadataTransaction>&&) @ 0x0000000015018de7
7. ./contrib/llvm-project/libcxx/include/__functional/function.h:716: ? @ 0x0000000015018c2f
8. ./build.release/./src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp:1067: DB::DiskObjectStorageTransaction::commit(std::variant<std::monostate, DB::MetaInKeeperCommitOptions<zkutil::ZooKeeper>, DB::MetaInKeeperCommitOptions<DB::ZooKeeperWithFaultInjection>> const&) @ 0x0000000015017e8f
9. ./build.release/./src/Disks/ObjectStorages/DiskObjectStorage.cpp:389: DB::DiskObjectStorage::createDirectories(String const&) @ 0x0000000015005775
10. ./build.release/./src/Storages/MergeTree/MergeTreeData.cpp:409: DB::MergeTreeData::initializeDirectoriesAndFormatVersion(String const&, bool, String const&, bool) @ 0x000000001903f500
11. ./build.release/./src/Storages/StorageMergeTree.cpp:170: DB::StorageMergeTree::StorageMergeTree(DB::StorageID const&, String const&, DB::StorageInMemoryMetadata const&, DB::LoadingStrictnessLevel, std::shared_ptr<DB::Context>, String const&, DB::MergeTreeData::MergingParams const&, std::unique_ptr<DB::MergeTreeSettings, std::default_delete<DB::MergeTreeSettings>>) @ 0x0000000019000a9c
12. ./contrib/llvm-project/libcxx/include/__memory/construct_at.h:41: std::shared_ptr<DB::StorageMergeTree> std::allocate_shared[abi:ne190107]<DB::StorageMergeTree, std::allocator<DB::StorageMergeTree>, DB::StorageID const&, String const&, DB::StorageInMemoryMetadata&, DB::LoadingStrictnessLevel const&, std::shared_ptr<DB::Context>&, String&, DB::MergeTreeData::MergingParams&, std::unique_ptr<DB::MergeTreeSettings, std::default_delete<DB::MergeTreeSettings>>, 0>(std::allocator<DB::StorageMergeTree> const&, DB::StorageID const&, String const&, DB::StorageInMemoryMetadata&, DB::LoadingStrictnessLevel const&, std::shared_ptr<DB::Context>&, String&, DB::MergeTreeData::MergingParams&, std::unique_ptr<DB::MergeTreeSettings, std::default_delete<DB::MergeTreeSettings>>&&) @ 0x000000001941a5f6
13. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:851: DB::create(DB::StorageFactory::Arguments const&) @ 0x000000001941720d
14. ./contrib/llvm-project/libcxx/include/__functional/function.h:716: ? @ 0x000000001897a6fb
15. ./build.release/./src/Interpreters/InterpreterCreateQuery.cpp:1951: DB::InterpreterCreateQuery::doCreateTable(DB::ASTCreateQuery&, DB::InterpreterCreateQuery::TableProperties const&, std::unique_ptr<DB::DDLGuard, std::default_delete<DB::DDLG26. start_thread @ 0x0000000000094ac3
27. __GI___clone3 @ 0x0000000000126850
 (version 25.9.1.1)
```

Leaving it as a no-op.